data/src/main/java/com/microsoft/azure/kusto/data/auth/TokenProviderBase.java

```diff
@@ -0,0 +1,104 @@
+        String scope = String.format("%s/%s", clusterUrl, ".default");
+        scopes = new HashSet<>();
+        scopes.add(scope);
+
+        aadAuthorityUrl = determineAadAuthorityUrl(authorityId);
+    }
+
+    private String determineAadAuthorityUrl(String authorityId) {
+        if (authorityId == null) {
+            authorityId = ORGANIZATION_URI_SUFFIX;
+        }
+
+        String authorityUrl;
+        String aadAuthorityFromEnv = System.getenv("AadAuthorityUri");
+        if (aadAuthorityFromEnv == null) {
+            authorityUrl = String.format("https://login.microsoftonline.com/%s/", authorityId);
+        } else {
+            authorityUrl = String.format("%s%s%s/", aadAuthorityFromEnv, aadAuthorityFromEnv.endsWith("/") ? "" : "/", authorityId);
+        }
+        return authorityUrl;
+    }
+
+    public String acquireAccessToken() throws DataServiceException, DataClientException {
+        IAuthenticationResult accessTokenResult = acquireAccessTokenSilently();
+        if (accessTokenResult == null) {
+            accessTokenResult = acquireNewAccessToken();
+        }
+        return accessTokenResult.accessToken();
+    }
+
+    protected IAuthenticationResult acquireAccessTokenSilently() throws DataServiceException, DataClientException {
+        try {
+            // TODO: MSAL docs say this would throw MsalUiRequiredException, but it doesn't in Java
+            return acquireAccessTokenSilentlyMsal();
+        } catch (MalformedURLException e) {
+            throw new DataClientException(clusterUrl, ERROR_INVALID_AUTHORITY_URL, e);
+        } catch (TimeoutException | ExecutionException e) {
+            return null; // Legitimate outcome, in which case a new token will be acquired
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            return null; // Legitimate outcome, in which case a new token will be acquired
+        }
+    }
+
+    protected IAuthenticationResult acquireAccessTokenSilentlyMsal() throws MalformedURLException, InterruptedException, ExecutionException, TimeoutException, DataServiceException {
+        throw new DataServiceException("Cannot obtain a refresh token for Authentication type '" + this.getClass().getSimpleName() + "'");
+    }
+
+    protected IAuthenticationResult acquireNewAccessToken() throws DataServiceException, DataClientException {
+        throw new DataServiceException("Cannot obtain a new token for Authentication type '" + this.getClass().getSimpleName() + "'");
+    }
+
+    SilentParameters getSilentParameters(Set<IAccount> accountSet) {
+        IAccount account = getAccount(accountSet);
+        if (account != null) {
+            return SilentParameters.builder(scopes).account(account).authorityUrl(aadAuthorityUrl).build();
+        }
+        return SilentParameters.builder(scopes).authorityUrl(aadAuthorityUrl).build();
+    }
+
+    IAccount getAccount(Set<IAccount> accountSet) {

```

Comment:
hint or only one for public

Reply:
See override in UserPromptTokenProvider.